added SimBA (Simple Black-box Adversarial Attacks)

kztakemoto opened the pull request:
# Description

## added SimBA
added code for SimBA (Simple Black-box Adversarial Attacks) implemented in Keras: `art/attacks/evasion/simba.py`. This is related to Issue #51 . 

modified UniversalPerturbation (`art/attacks/evasion/universal_perturbation.py `) to use SimBA in UniversalPerturbation as an attacker.

## added a targeted attack version of UniversalPerturbation
added code for targeted attack version of UniversalPerturbation (TargetedUniversalPerturbation): `art/attacks/evasion/targeted_universal_perturbation.py` (see also https://github.com/hkthirano/targeted_UAP_CIFAR10). Note that FGSM and SimBA are only available as attackers in this implementation. This is related to Issue #393 .

## modified UniversalPerturbation
to be able to give associated correct labels (i.e., y) because the current version of UniversalPerturbation only use the model predictions as the correct labels. This is related to Issue #393 .

Fixes # (issue)

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

I have tested this implementation with some examples: https://github.com/kztakemoto/SimBA-Keras

**Test Configuration**:
- OS: macOS Catalina version 10.15.5  
- Python version 3.7.7
- ART version: 1.1.0
- Keras (v 2.3.1)

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

beat-buesser commented:
Hi @kztakemoto Thank you very much for this pull request implementing your new evasion attacks! We will start a review of this pull request in the coming days.
I'm changing the target branch for this pull request to `dev_1.4.0` which is the development branch for the next possible release of ART.
Please let us know anytime if you have questions!

beat-buesser commented:
@kztakemoto There are some merge conflicts in this pull request. Could you please pull and merge our branch `dev_1.4.0` into your branch `kztakemoto:add-simba` and solve the merge conflicts? Please let us know if you have any questions.

kztakemoto commented:
@beat-buesser Thanks. I resolved the conflicts

beat-buesser commented on art/attacks/__init__.py:
```diff
@@ -2,4 +2,4 @@
 """
 from art.attacks.attack import Attack, EvasionAttack, PoisoningAttack, PoisoningAttackBlackBox, PoisoningAttackWhiteBox
-from art.attacks.attack import ExtractionAttack, InferenceAttack, AttributeInferenceAttack
+from art.attacks.attack import ExtractionAttack, InferenceAttack, AttributeInferenceAttack
```
Please don't remove the newline character at the end of this line.
```suggestion
from art.attacks.attack import ExtractionAttack, InferenceAttack, AttributeInferenceAttack

```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2018
```
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```

beat-buesser commented on art/estimators/classification/pytorch.py:
```diff
@@ -261,12 +261,12 @@ def fit_generator(self, generator: "DataGenerator", nb_epochs: int = 20, **kwarg
                 for i_batch, o_batch in generator.iterator:
                     if isinstance(i_batch, np.ndarray):
-                        i_batch = torch.from_numpy(i_batch).to(self._device)
+                        i_batch = torch.from_numpy(i_batch).type(torch.FloatTensor).to(self._device)
```
What is the reason for this and similar additions in lines 269, 318, 396, 404, and 496?

kztakemoto replied:
Our modifications are neglectable. These were just for temporary measures. I have to push the corrected version (i.e., the original version in IBM:dev_1.4.0)?

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+import logging
+
+import numpy as np
+from scipy.fftpack import dct, idct
```
The imported `dct` seems not to be used in this file.
```suggestion
from scipy.fftpack import idct
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+from art.attacks.attack import EvasionAttack
+from art.classifiers.classifier import ClassifierGradients
+from art.config import ART_NUMPY_DTYPE
+from art.utils import compute_success
```
`compute_success` seems not to be used in this file
```suggestion
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+        :param batch_size: Batch size (but, batch process unavailable in this implementation)
+        :type batch_size: `int`
+        """
+        super(SimBA, self).__init__(classifier=classifier)
```
The keyword change to `estimator` in ART 1.3.
```suggestion
        super(SimBA, self).__init__(estimator=classifier)
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+
+class SimBA(EvasionAttack):
+    attack_params = EvasionAttack.attack_params + ['attack', 'max_iter', 'epsilon', 'order', 'freq_dim', 'stride', 'targeted', 'batch_size',]
+
```
```suggestion

    _estimator_requirements = (BaseEstimator, ClassGradientsMixin)

```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+        :type batch_size: `int`
+        """
+        super(SimBA, self).__init__(classifier=classifier)
+        if not isinstance(classifier, ClassifierGradients):
+            raise (TypeError('For `' + self.__class__.__name__ + '` classifier must be an instance of '
+                             '`art.classifiers.classifier.ClassifierGradients`, the provided classifier is instance of '
+                             + str(classifier.__class__.__bases__) + '. '
+                             ' The classifier needs to be a Neural Network and provide gradients.'))
```
This check has been replaced by `_estimator_requirements`

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+
+        return x
+
+    def set_params(self, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks before saving them as attributes.
+
+        :param attack: attack type: pixel (px) or DCT (dct) attacks
+        :type attack: `str`
+        :param max_iter: The maximum number of iterations.
+        :type max_iter: `int`
+        :param epsilon: Overshoot parameter.
+        :type epsilon: `float`
+        :param order: order of pixel attacks
+        :type order: `str`
+        :param freq_dim: dimensionality of 2D frequency space (DCT).
+        :type freq_dim: `int`
+        :param stride: stride for block order (DCT).
+        :type stride: `int`
+        :param targeted: targeted attacks
+        :type targeted: `bool`
+        :param batch_size: Batch size (but, batch process unavailable in this implementation)
+        :type batch_size: `int`
+        """
+        # Save attack-specific parameters
+        super(SimBA, self).set_params(**kwargs)
```
```suggestion
    def _check_params(self) -> None:
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+        
+        if not isinstance(self.targeted, (int)) or (self.targeted != 0 and self.targeted != 1):
+            raise ValueError('`targeted` has to be a logical value.')
+
+        return True
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+                             + str(classifier.__class__.__bases__) + '. '
+                             ' The classifier needs to be a Neural Network and provide gradients.'))
+
+        params = {'attack': attack, 'max_iter': max_iter, 'epsilon': epsilon, 'order': order, 'freq_dim': freq_dim, 'stride': stride, 'targeted': targeted ,'batch_size': batch_size}
+        self.set_params(**params)
```
```suggestion
        self.attack = attack
        self.max_iter = max_iter
        self.epsilon = epsilon
        self.order = order
        self.freq_dim = freq_dim
        self.stride = stride
        self.targeted = targeted
        self.batch_size = batch_size
        self._check_params()
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+class SimBA(EvasionAttack):
+    attack_params = EvasionAttack.attack_params + ['attack', 'max_iter', 'epsilon', 'order', 'freq_dim', 'stride', 'targeted', 'batch_size',]
+
+    def __init__(self, classifier, attack='dct', max_iter=3000, order='random', epsilon=0.1, freq_dim=4, stride=1, targeted=False, batch_size=1):
+        """
+        Create a SimBA (dct) attack instance.
+
+        :param classifier: A trained classifier.
+        :type classifier: :class:`.Classifier`
+        :param attack: attack type: pixel (px) or DCT (dct) attacks
+        :type attack: `str`
+        :param max_iter: The maximum number of iterations.
+        :type max_iter: `int`
+        :param epsilon: Overshoot parameter.
+        :type epsilon: `float`
+        :param order: order of pixel attacks: random or diagonal (diag)
+        :type order: `str`
+        :param freq_dim: dimensionality of 2D frequency space (DCT).
+        :type freq_dim: `int`
+        :param stride: stride for block order (DCT).
+        :type stride: `int`
+        :param targeted: perform targeted attack
+        :type targeted: `bool`
+        :param batch_size: Batch size (but, batch process unavailable in this implementation)
+        :type batch_size: `int`
+        """
```
```suggestion
    def __init__(self,
                        classifier: ClassifierGradients,
                        attack: str = 'dct',
                        max_iter: int = 3000,
                        order: str = 'random',
                        epsilon: float = 0.1,
                        freq_dim: int = 4,
                        stride: int = 1,
                        targeted: bool = False,
                        batch_size: int = 1):
        """
        Create a SimBA (dct) attack instance.

        :param classifier: A trained classifier.
        :type classifier: :class:`.Classifier`
        :param attack: attack type: pixel (px) or DCT (dct) attacks
        :type attack: `str`
        :param max_iter: The maximum number of iterations.
        :type max_iter: `int`
        :param epsilon: Overshoot parameter.
        :type epsilon: `float`
        :param order: order of pixel attacks: random or diagonal (diag)
        :type order: `str`
        :param freq_dim: dimensionality of 2D frequency space (DCT).
        :type freq_dim: `int`
        :param stride: stride for block order (DCT).
        :type stride: `int`
        :param targeted: perform targeted attack
        :type targeted: `bool`
        :param batch_size: Batch size (but, batch process unavailable in this implementation)
        :type batch_size: `int`
        """
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+
+| Paper link: https://arxiv.org/abs/1905.07121
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
```
```suggestion
from __future__ import (absolute_import, division, print_function, unicode_literals)
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+        :rtype: `np.ndarray`
+        """
+        x = x.astype(ART_NUMPY_DTYPE)
+        preds = self.classifier.predict(x, batch_size=self.batch_size)
```
```suggestion
        preds = self.estimator.predict(x, batch_size=self.batch_size)
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+        params = {'attack': attack, 'max_iter': max_iter, 'epsilon': epsilon, 'order': order, 'freq_dim': freq_dim, 'stride': stride, 'targeted': targeted ,'batch_size': batch_size}
+        self.set_params(**params)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs to be attacked.
+        :type x: `np.ndarray`
+        :param y: An array with the original labels to be predicted.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
```
```suggestion
    def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.ndarray:
        """
        Generate adversarial samples and return them in an array.

        :param x: An array with the original inputs to be attacked.
        :param y: An array with the original labels to be predicted.
        :return: An array holding the adversarial examples.
        """
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+
+        clip_min = -np.inf
+        clip_max = np.inf 
+        if hasattr(self.classifier, 'clip_values') and self.classifier.clip_values is not None:
```
`clip_values` is now a property.
```suggestion
        if self.classifier.clip_values is not None:
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+
+        return True
+
+    def _block_order(self, img_size, channels, initial_size=2, stride=1):
```
Please add a docstring and typing information to this method.

kztakemoto replied:
added a docstring and typing information to this method.

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+        return order.transpose(1,2,0).reshape(1, -1).squeeze().argsort()
+
+    # applies IDCT to each block of size block_size
+    def _block_idct(self, x, block_size=8, masked=False, ratio=0.5):
```
Please add a docstring and typing information to this method.

kztakemoto replied:
added a docstring and typing information to this method.

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,319 @@
+                z[:, :, (i * block_size):((i + 1) * block_size), (j * block_size):((j + 1) * block_size)] = idct(idct(submat, axis=3, norm='ortho'), axis=2, norm='ortho')
+        return z.transpose(0,2,3,1)
+    
+    def diagonal_order(self, image_size, channels):
```
Please add a docstring and typing information to this method.

kztakemoto replied:
added a docstring and typing information to this method.

beat-buesser commented on art/attacks/evasion/universal_perturbation.py:
```diff
@@ -42,6 +42,19 @@
 
 logger = logging.getLogger(__name__)
 
+#####
+import numpy as np
+import matplotlib
+matplotlib.use("Agg")
+import matplotlib.pyplot as plt
+import japanize_matplotlib
+
+import torch
+import logging
+import os.path
+
+import numpy as np
+
```
These import seem redundant.
```suggestion
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2018
```
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+                    'simba': 'art.attacks.evasion.simba.SimBA'
+                    }
+    attack_params = EvasionAttack.attack_params + ['attacker', 'attacker_params', 'delta', 'max_iter', 'eps', 'norm']
+
```
```suggestion

    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin, ClassGradientsMixin, LossGradientsMixin)

```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+        :type norm: `int`
+        """
+        super(TargetedUniversalPerturbation, self).__init__(classifier)
+        if not isinstance(classifier, ClassifierNeuralNetwork) or not isinstance(classifier, ClassifierGradients):
+            raise (TypeError('For `' + self.__class__.__name__ + '` classifier must be an instance of '
+                             '`art.classifiers.classifier.ClassifierNeuralNetwork` and '
+                             '`art.classifiers.classifier.ClassifierGradients`, the provided classifier is instance of '
+                             + str(classifier.__class__.__bases__) + '. '
+                             ' The classifier needs to be a Neural Network and provide gradients.'))
```
This has been replaced by `_estimator_requirements`.
```suggestion
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+        :param norm: The norm of the adversarial perturbation. Possible values: np.inf, 2
+        :type norm: `int`
+        """
+        super(TargetedUniversalPerturbation, self).__init__(classifier)
```
```suggestion
        super(TargetedUniversalPerturbation, self).__init__(estimator=classifier)
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+                    }
+    attack_params = EvasionAttack.attack_params + ['attacker', 'attacker_params', 'delta', 'max_iter', 'eps', 'norm']
+
+    def __init__(self, classifier, attacker='fgsm', attacker_params=None, delta=0.2, max_iter=20, eps=10.0,
+                 norm=np.inf):
+        """
+        :param classifier: A trained classifier.
+        :type classifier: :class:`.Classifier`
+        :param attacker: Adversarial attack name. Default is 'deepfool'. Supported names: 'fgsm'.
+        :type attacker: `str`
+        :param attacker_params: Parameters specific to the adversarial attack. If this parameter is not specified,
+                                the default parameters of the chosen attack will be used.
+        :type attacker_params: `dict`
+        :param delta: desired accuracy
+        :type delta: `float`
+        :param max_iter: The maximum number of iterations for computing universal perturbation.
+        :type max_iter: `int`
+        :param eps: Attack step size (input variation)
+        :type eps: `float`
+        :param norm: The norm of the adversarial perturbation. Possible values: np.inf, 2
+        :type norm: `int`
+        """
```
```suggestion
    def __init__(self,
                        classifier: ClassifierGradients,
                        attacker: str = 'fgsm',
                        attacker_params: Dict[str, Any] = None,
                        delta: float = 0.2,
                        max_iter: int = 20,
                        eps: float = 10.0,
                        norm: int = np.inf):
        """
        :param classifier: A trained classifier.
        :param attacker: Adversarial attack name. Default is 'deepfool'. Supported names: 'fgsm'.
        :param attacker_params: Parameters specific to the adversarial attack. If this parameter is not specified,
                                the default parameters of the chosen attack will be used.
        :param delta: desired accuracy
        :param max_iter: The maximum number of iterations for computing universal perturbation.
        :param eps: Attack step size (input variation)
        :param norm: The norm of the adversarial perturbation. Possible values: np.inf, 2
        """
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+                             + str(classifier.__class__.__bases__) + '. '
+                             ' The classifier needs to be a Neural Network and provide gradients.'))
+
+        kwargs = {'attacker': attacker,
+                  'attacker_params': attacker_params,
+                  'delta': delta,
+                  'max_iter': max_iter,
+                  'eps': eps,
+                  'norm': norm
+                  }
+        self.set_params(**kwargs)
```
```suggestion
        self.attacker = attacker
        self.attacker_params = attacker_params
        self.delta = delta
        self.max_iter = max_iter
        self.eps = eps
        self.norm = norm
        self._check_params()
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+                  }
+        self.set_params(**kwargs)
+
+    def generate(self, x, y, **kwargs):
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs.
+        :type x: `np.ndarray`
+        :param y: An array with the targeted labels.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
```
```suggestion
    def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.ndarray:
        """
        Generate adversarial samples and return them in an array.

        :param x: An array with the original inputs.
        :param y: An array with the targeted labels.
        :return: An array holding the adversarial examples.
        """
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+
+        return x_adv
+
+    def set_params(self, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks before saving them as attributes.
+
+        :param attacker: Adversarial attack name. Default is 'deepfool'. Supported names: 'fgsm'.
+        :type attacker: `str`
+        :param attacker_params: Parameters specific to the adversarial attack.
+        :type attacker_params: `dict`
+        :param delta: desired accuracy
+        :type delta: `float`
+        :param max_iter: The maximum number of iterations for computing universal perturbation.
+        :type max_iter: `int`
+        :param eps: Attack step size (input variation)
+        :type eps: `float`
+        :param norm: Order of the norm. Possible values: np.inf, 2 (default is np.inf)
+        :type norm: `int`
+        """
+        super(TargetedUniversalPerturbation, self).set_params(**kwargs)
```
```suggestion
    def _check_params(self) -> None:
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+        if not isinstance(self.eps, (float, int)) or self.eps <= 0:
+            raise ValueError("The eps coefficient must be a positive float.")
+
+        return True
+
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+
+        return True
+
+    def _get_attack(self, a_name, params=None):
+        """
+        Get an attack object from its name.
+
+        :param a_name: attack name.
+        :type a_name: `str`
+        :param params: attack params.
+        :type params: `dict`
+        :return: attack object
+        :rtype: `object`
+        """
```
```suggestion
    def _get_attack(self, a_name: str, params: Optional[Dict[str, Any]] = None) - > EvasionAttack:
        """
        Get an attack object from its name.

        :param a_name: attack name.
        :param params: attack params.
        :return: attack object
        """
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+            raise NotImplementedError("{} attack not supported".format(a_name))
+
+    @staticmethod
+    def _get_class(class_name):
+        """
+        Get a class module from its name.
+
+        :param class_name: Full name of a class.
+        :type class_name: `str`
+        :return: The class `module`.
+        :rtype: `module`
+        """
```
```suggestion
    def _get_class(class_name: str) -> types.ModuleType:
        """
        Get a class module from its name.

        :param class_name: Full name of a class.
        :return: The class `module`.
        """
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,226 @@
+        :rtype: `module`
+        """
+        sub_mods = class_name.split(".")
+        print(class_name)
```
```suggestion
```

kztakemoto commented:
Hi @beat-buesser I really appreciate your review. I

- confirmed and accepted all your suggestions.
- added docstring and typing information to the methods.
- further modified our code (simba.py and targeted_universal_perturbation.py) to run successfully, after accepting all suggestions.
- confirmed that our code run successfully.
- added unittests for SimBA and TargetedUniversalPerturbation (but only for MNIST)

Note: the current version of SimBA can run only in Keras and TensorFlow (not in PyTorch).

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,339 @@
+from __future__ import (absolute_import, division, print_function, unicode_literals)
+
+import logging
+import types
+from typing import Any, Dict, Optional, Union
```
```suggestion
from typing import Optional
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,339 @@
+
+| Paper link: https://arxiv.org/abs/1905.07121
+"""
+from __future__ import (absolute_import, division, print_function, unicode_literals)
```
```suggestion
from __future__ import absolute_import, division, print_function, unicode_literals
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,198 @@
+import logging
+import random
+import types
+from typing import Any, Dict, Optional, Union
```
```suggestion
from typing import Any, Dict, Optional
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,198 @@
+
+import numpy as np
+
+from art.classifiers.classifier import ClassifierNeuralNetwork, ClassifierGradients
+from art.attacks.attack import EvasionAttack
+from art.estimators.estimator import BaseEstimator, NeuralNetworkMixin, LossGradientsMixin
+from art.estimators.classification.classifier import (
+    ClassGradientsMixin,
+)
```
```suggestion
from art.attacks.attack import EvasionAttack
from art.estimators.estimator import BaseEstimator, NeuralNetworkMixin, LossGradientsMixin
from art.estimators.classification.classifier import (
    ClassifierGradients,
    ClassGradientsMixin,
)
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,339 @@
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the black-box attack `simba`.
```
```suggestion
This module implements the black-box attack `SimBA`.
```

beat-buesser commented on art/estimators/classification/pytorch.py:
```diff
@@ -401,7 +401,7 @@ def loss_gradient(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
         # Compute the gradient and return
         model_outputs = self._model(inputs_t)
-        loss = self._loss(model_outputs[-1], labels_t)
+        loss = self._loss(model_outputs[-1], labels_t.long())
```
Why is this change required?

kztakemoto replied:
Our modification is neglectable. This was just a temporary measure. I pushed the corrected version (i.e., the original version in IBM:dev_1.4.0).

beat-buesser commented on tests/attacks/test_targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,167 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2018
```
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```

beat-buesser commented on tests/attacks/test_simba.py:
```diff
@@ -0,0 +1,152 @@
+    Note: SimBA runs only in Keras and TensorFlow (not in PyTorch)
+    This is due to the channel first format in PyTorch.
+
+    | SimBA Paper link:
+        https://arxiv.org/abs/1905.07121
```
```suggestion
    | Paper link: https://arxiv.org/abs/1905.07121
```

beat-buesser commented on tests/attacks/test_targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,167 @@
+
+    This module tests the Targeted Universal Perturbation.
+
+    | Targeted Universal Perturbation Paper:
+       Hirano H and Takemoto K (2020) Simple iterative method for generating targeted universal adversarial perturbations. in Proceedings of 25th International Symposium on Artificial Life and Robotics (AROB 25th 2020), pp. 426-430.
+        (arXiv link: https://arxiv.org/abs/1911.06502)
```
```suggestion
    | Paper link: https://arxiv.org/abs/1911.06502)
```

beat-buesser commented on art/attacks/evasion/targeted_universal_perturbation.py:
```diff
@@ -0,0 +1,198 @@
+        """
+        try:
+            attack_class = self._get_class(self.attacks_dict[a_name])
+            a_instance = attack_class(self.estimator)
```
```suggestion
            a_instance = attack_class(self.estimator)  # type: ignore
```

beat-buesser commented:
Hi @kztakemoto Thank you very much for your patience and fixing all requests. I have found a few more change requests mainly about fixing imports and formats, could you please take a look?

Once this PR is merged we will list your paper in our wiki page: https://github.com/IBM/adversarial-robustness-toolbox/wiki/ART-Attacks

To make the new unittests run on Travis CI could you please add your 2 tests to the file `run_tests.sh` after line 57?

To fix PEP8 format and line lengths could you please run `black` on all modified files using a commands like `black --line-length 120 art/attacks/evasion/simba.py`? You can install `black` with `pip install black`.

Why does `SimBA` not work for PyTorch? Is it because of the channels first format required for PyTorch?
If yes, would it be possible to make SimBA work for PyTorch by swapping axes of `x` in method `generate` similar as for example in https://github.com/IBM/adversarial-robustness-toolbox/blob/307f7c9ec255c9c3594a9e57c82a2b9064c69843/art/attacks/evasion/wasserstein.py#L603 and revert it later like in https://github.com/IBM/adversarial-robustness-toolbox/blob/307f7c9ec255c9c3594a9e57c82a2b9064c69843/art/attacks/evasion/wasserstein.py#L622?

kztakemoto commented:
Hi @beat-buesser Thanks for your review again. According to your suggestions, I
- added the 2 tests to the file `run_tests.sh` after line 57.
- modified `art/attacks/evasion/simba.py` to run in PyTorch (thanks for the information about `self.estimator.channels_first`!).
- added PyTorch tests into `tests/attacks/test_simba.py`.
- run on all modified files to fix PEP8 format and line lengths as follows: `black --line-length 120 [path-to-file]`.
- confirmed that all modified files run correctly.

We are really happy that you are planning to list our paper in your wiki page!

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,377 @@
+        :param x: An array with the inputs to be attacked.
+        :param block_size: block size for DCT attacks.
+        :param masked: use the mask.
+        :param ratio: Ratio of the lowest frequency directions in order to make the adversarial perturbation in the low frequency space.
```
```suggestion
        :param ratio: Ratio of the lowest frequency directions in order to make the adversarial perturbation in the low
                      frequency space.
```

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,377 @@
+                tmp_indices = self._block_order(x.shape[2], nb_channels, initial_size=self.freq_dim, stride=self.stride)
+                indices = np.hstack((indices, tmp_indices))[: self.max_iter]
+                indices_size = len(indices)
+            trans = lambda z: self._block_idct(z, block_size=x.shape[2])
```
```suggestion

            def trans(z): return self._block_idct(z, block_size=x.shape[2])
```

beat-buesser commented:
Hi @kztakemoto Thank you very much!

The new tests are now successfully running on Travis and the LGTM analysis does not raise any alerts.

The PEP8 check has found two more warnings that `black` does not fix, one about a too long line in a docstring and one for a lambda assignment. I have made suggestions above on how to resolve them.

I think after these two changes all the checks should be passing.

kztakemoto commented:
@beat-buesser Thanks again. All your suggestions were accepted. We hope all the checks are passing.

beat-buesser commented on art/attacks/evasion/simba.py:
```diff
@@ -0,0 +1,379 @@
+                indices = np.hstack((indices, tmp_indices))[: self.max_iter]
+                indices_size = len(indices)
+
+            def trans(z): return self._block_idct(z, block_size=x.shape[2])
```
```suggestion
            def trans(z):
                return self._block_idct(z, block_size=x.shape[2])
```

beat-buesser commented:
@kztakemoto I'm sorry, my last suggestion lead to another PEP8 warning, this last suggestion above fixes it.

kztakemoto commented:
@beat-buesser thanks for your kind cooperation!!

beat-buesser commented:
Hi @kztakemoto Thank you very much for your contribution which is adding your Targeted Universal Perturbation attack and the Simple Black-box Adversarial Attacks attack to ART and for your work and patience during the review process!

kztakemoto commented:
@beat-buesser Thank you very much. We are really happy to contribute the development of ART!
